finesse/hardware/em27_opus.py

```diff
@@ -0,0 +1,147 @@
+from pubsub import pub
+from PySide6.QtCore import QObject, QThread, Signal, Slot
+
+from ..config import OPUS_IP
+
+STATUS_FILENAME = "stat.htm"
+COMMAND_FILENAME = "cmd.htm"
+
+
+class OPUSError(Exception):
+    """Indicates that an error occurred while communicating with the OPUS program."""
+
+
+class OPUSRequester(QThread):
+    """Interface for making HTTP requests on a background thread."""
+
+    request_complete = Signal(requests.Request, str)
+    request_error = Signal(BaseException)
+
+    def __init__(self, timeout: float) -> None:
+        """Create a new OPUSRequester."""
+        super().__init__()
+        self.timeout = timeout
+        self.moveToThread(self)
+
+    @Slot()
+    def make_request(self, filename: str, topic: str):
+        """Make a request to the OPUS program.
+
+        Args:
+            filename: The final part of the path on the HTTP server
+            topic: The topic on which to publish the response
+        """
+        try:
+            url = f"http://{OPUS_IP}/opusrs/{filename}"
+            response = requests.get(url, timeout=self.timeout)
+            self.request_complete.emit(response, topic)
+        except Exception as e:
+            self.request_error.emit(e)
+
+
+class OPUSInterface(QObject):
+    """Interface for communicating with the OPUS program.
+
+    HTTP requests are handled on a background thread.
+    """
+
+    submit_request = Signal(str, str)
+    """Signal indicating that an HTTP request should be made."""
+
+    def __init__(self, timeout: float = 3.0) -> None:
+        """Create a new OPUSInterface.
+
+        Args:
+            timeout: Amount of time before request times out (seconds)
+        """
+        super().__init__()
+
+        self.requester = OPUSRequester(timeout)
+        """For running HTTP requests in the background."""
+        self.requester.request_complete.connect(self._parse_response)
+        self.requester.request_error.connect(self._error_occurred)
+
+        # Set up a signal for communicating between threads
+        self.submit_request.connect(self.requester.make_request)
+
+        # Subscribe to requests coming from the GUI
+        pub.subscribe(self.request_status, "opus.status.request")
+        pub.subscribe(self.request_command, "opus.command.request")
+
+        # Start processing requests
+        self.requester.start()
+
+    def __del__(self) -> None:
+        """Stop the background request thread."""
+        self.requester.quit()
+        self.requester.wait()
+
+    def _parse_response(self, response: requests.Response, topic: str) -> None:
+        try:
+            if response.status_code != 200:
+                raise OPUSError(f"HTTP status code {response.status_code}")
+
+            status: Optional[int] = None
+            text = ""
+            errcode: Optional[int] = None
+            errtext = ""
+            soup = BeautifulSoup(response.content, "html.parser")
+            for td in soup.find_all("td"):
```

Comment:
I guess as long as the html doesn't decide to capitalise all its tags this should be fine!

Reply:
Fingers crossed! If it does we can always revisit this.